### PR TITLE
Emit JSON schema and BigQuery schema

### PR DESCRIPTION
**Pull request by duncanjbrown**
When setting up a new project using `dfe-analytics`, it's necessary to create a schema in BigQuery that's compatible with the schema in this project.

BigQuery uses its own schema format, so we need to map the JSON schema to that format. The BigQuery schema contains less information than the JSON schema (e.g. it lacks enums, `additionalProperties` etc). We keep the JSON schema as the source of truth for testing, and generate a BigQuery schema from it.

This PR adds two new rake tasks: `dfe:analytics:schema` and `dfe:analytics:big_query_schema`. The BigQuery schema is suitable for copy-pasting into the BigQuery web interface. It could also be used to generate a table via the `bq` CLI tool.

### Changes to the schema!

During this process I noticed that the event-schema.json was not the same as the BigQuery schema (for Apply, at least) and I amended it (relaxing requirements, so not breaking changes)

### CODE REVIEW

**Review by tomas-stefano — APPROVED** [review]

**Comment by misaka** [comment]
Comparing what comes out of the `bq` cmdline and what this generates is interesting. There a couple of small differences:

## JSON generated by this task

```json
[
  {
    "name": "occurred_at",
    "type": "TIMESTAMP",
    "mode": "REQUIRED"
  },
  {
    "name": "event_type",
    "type": "STRING",
    "mode": "REQUIRED"
  },
  {
    "name": "environment",
    "type": "STRING",
    "mode": "REQUIRED"
  },
  {
    "name": "namespace",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "user_id",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_uuid",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_method",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_path",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_user_agent",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_referer",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_query",
    "type": "RECORD",
    "mode": "REPEATED",
    "fields": [
      {
        "name": "key",
        "type": "STRING",
        "mode": "REQUIRED"
      },
      {
        "name": "value",
        "type": "STRING",
        "mode": "REPEATED"
      }
    ]
  },
  {
    "name": "response_content_type",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "response_status",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "data",
    "type": "RECORD",
    "mode": "REPEATED",
    "fields": [
      {
        "name": "key",
        "type": "STRING",
        "mode": "REQUIRED"
      },
      {
        "name": "value",
        "type": "STRING",
        "mode": "REPEATED"
      }
    ]
  },
  {
    "name": "entity_table_name",
    "type": "STRING"
  },
  {
    "name": "event_tags",
    "type": "STRING",
    "mode": "REPEATED"
  },
  {
    "name": "anonymised_user_agent_and_ip",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "row_id",
    "type": "STRING",
    "mode": "NULLABLE"
  }
]
```

## JSON generated by bq

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "environment": {
      "type": "string"
    },
    "occurred_at": {
      "type": "string",
      "format": "date-time"
    },
    "event_type": {
      "type": "string",
      "enum": [
        "web_request",
        "create_entity",
        "update_entity",
        "delete_entity",
        "import_entity"
      ]
    },
    "entity_table_name": {
      "type": "string"
    },
    "request_uuid": {
      "type": "string"
    },
    "request_user_agent": {
      "type": "string"
    },
    "request_method": {
      "type": "string"
    },
    "request_path": {
      "type": "string"
    },
    "request_query": {
      "type": "array",
      "items": {
        "type": "object",
        "required": [
          "key",
          "value"
        ]
      }
    },
    "request_referer": {
      "type": "string"
    },
    "anonymised_user_agent_and_ip": {
      "type": "string"
    },
    "response_content_type": {
      "type": "string"
    },
    "response_status": {
      "type": "string"
    },
    "namespace": {
      "type": "string"
    },
    "user_id": {
      "type": "string"
    },
    "data": {
      "type": "array",
      "items": [
        {
          "type": "object",
          "required": [
            "key",
            "value"
          ]
        }
      ]
    }
  },
  "additionalProperties": false,
  "required": [
    "environment",
    "occurred_at",
    "event_type"
  ]
}
```

**Comment by misaka** [comment]
Comparing what comes out of the `bq` cmdline and what this generates is interesting. There a couple of small differences:

## JSON generated by this task

```json
[
  {
    "name": "occurred_at",
    "type": "TIMESTAMP",
    "mode": "REQUIRED"
  },
  {
    "name": "event_type",
    "type": "STRING",
    "mode": "REQUIRED"
  },
  {
    "name": "environment",
    "type": "STRING",
    "mode": "REQUIRED"
  },
  {
    "name": "namespace",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "user_id",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_uuid",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_method",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_path",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_user_agent",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_referer",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "request_query",
    "type": "RECORD",
    "mode": "REPEATED",
    "fields": [
      {
        "name": "key",
        "type": "STRING",
        "mode": "REQUIRED"
      },
      {
        "name": "value",
        "type": "STRING",
        "mode": "REPEATED"
      }
    ]
  },
  {
    "name": "response_content_type",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "response_status",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "data",
    "type": "RECORD",
    "mode": "REPEATED",
    "fields": [
      {
        "name": "key",
        "type": "STRING",
        "mode": "REQUIRED"
      },
      {
        "name": "value",
        "type": "STRING",
        "mode": "REPEATED"
      }
    ]
  },
  {
    "name": "entity_table_name",
    "type": "STRING"
  },
  {
    "name": "event_tags",
    "type": "STRING",
    "mode": "REPEATED"
  },
  {
    "name": "anonymised_user_agent_and_ip",
    "type": "STRING",
    "mode": "NULLABLE"
  },
  {
    "name": "row_id",
    "type": "STRING",
    "mode": "NULLABLE"
  }
]
```

## JSON generated by bq

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "environment": {
      "type": "string"
    },
    "occurred_at": {
      "type": "string",
      "format": "date-time"
    },
    "event_type": {
      "type": "string",
      "enum": [
        "web_request",
        "create_entity",
        "update_entity",
        "delete_entity",
        "import_entity"
      ]
    },
    "entity_table_name": {
      "type": "string"
    },
    "request_uuid": {
      "type": "string"
    },
    "request_user_agent": {
      "type": "string"
    },
    "request_method": {
      "type": "string"
    },
    "request_path": {
      "type": "string"
    },
    "request_query": {
      "type": "array",
      "items": {
        "type": "object",
        "required": [
          "key",
          "value"
        ]
      }
    },
    "request_referer": {
      "type": "string"
    },
    "anonymised_user_agent_and_ip": {
      "type": "string"
    },
    "response_content_type": {
      "type": "string"
    },
    "response_status": {
      "type": "string"
    },
    "namespace": {
      "type": "string"
    },
    "user_id": {
      "type": "string"
    },
    "data": {
      "type": "array",
      "items": [
        {
          "type": "object",
          "required": [
            "key",
            "value"
          ]
        }
      ]
    }
  },
  "additionalProperties": false,
  "required": [
    "environment",
    "occurred_at",
    "event_type"
  ]
}
```

**Comment by misaka** [comment]
Putting comments on the differences in separate comments.

>     "event_type": {
>       "type": "string",
>       "enum": [
>         "web_request",
>         "create_entity",
>         "update_entity",
>         "delete_entity",
>         "import_entity"
>       ]
>     },

Sounds like this might be worth adding this to the JSON we generate.

**Comment by misaka** [comment]
Putting comments on the differences in separate comments.

>     "event_type": {
>       "type": "string",
>       "enum": [
>         "web_request",
>         "create_entity",
>         "update_entity",
>         "delete_entity",
>         "import_entity"
>       ]
>     },

Sounds like this might be worth adding this to the JSON we generate.

**Comment by misaka** [comment]
BQ version:
>   {
>     "name": "occurred_at",
>     "type": "TIMESTAMP",
>     "mode": "REQUIRED"
>   },

Our version:
>     "occurred_at": {
>       "type": "string",
>       "format": "date-time"
>     },

Should we be using `timestamp` here?

**Comment by misaka** [comment]
BQ version:
>   {
>     "name": "occurred_at",
>     "type": "TIMESTAMP",
>     "mode": "REQUIRED"
>   },

Our version:
>     "occurred_at": {
>       "type": "string",
>       "format": "date-time"
>     },

Should we be using `timestamp` here?

**Comment by misaka** [comment]
I've just tried to load the generated JSON schema using the `bq` cmdline and it's not happy with the formatting:

```
$ bq mk register_events_test.events_test events_def.json
BigQuery error in mk operation: Error in "events_def.json": Table schemas must be specified as JSON lists.
```

**Comment by misaka** [comment]
I've just tried to load the generated JSON schema using the `bq` cmdline and it's not happy with the formatting:

```
$ bq mk register_events_test.events_test events_def.json
BigQuery error in mk operation: Error in "events_def.json": Table schemas must be specified as JSON lists.
```

File: lib/dfe/analytics/event_schema.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
+module DfE
+  module Analytics
+    class EventSchema
```
Comment (misaka) [inline lib/dfe/analytics/event_schema.rb:3]:
```suggestion
    module EventSchema
```

Why is this a class and not a module?

**Comment by misaka** [comment]
I can't for the life of me understand why Rubocop is singling out `lib/dfe/analytics/event_schema.rb` for not having top-level documentation. Other classes and modules don't have top-level docs. I event tried copying over the entire contents of `load_entities.rb` into `event_schema.rb`, and Rubocop complains about `event_schema.rb` and not `load_entities.rb`! 🤯

```
$ be rubocop lib/dfe/analytics/load_entities.rb lib/dfe/analytics/event_schema.rb
Inspecting 2 files
.C

Offenses:

lib/dfe/analytics/event_schema.rb:5:5: C: Style/Documentation: Missing top-level documentation comment for class DfE::Analytics::LoadEntities.
    class LoadEntities
    ^^^^^^^^^^^^^^^^^^

2 files inspected, 1 offense detected
```

**Pull request merged**